### ISSUE TITLE
Polar Coordinates

### ISSUE THREAD

**Issue by Xaeroxe** [amethyst/amethyst#392]
These are super useful for describing orbital motions, it would be nice to see these integrated into either Amethyst itself or an upstream math library.

**Comment by Binero** [comment]
This is what I proposed in #342:

`Transform` works for all coordinate systems. If we make `LocalTransform` also just a wrapper around a model matrix, we can have one system which generates the `Transform` of every entity regardless of its coordinate system.

We could further specify a new component `CartesianCoordinates` and a new component `PolarCoordinates`, and a system to convert both of those into `LocalTransform` respectively. If someone uses a more exotic coordinate system, they could likewise create a new component to represent it, and a system to go with it.

**Comment by Xaeroxe** [comment]
So, I'd actually like to call for a simplification of our Transform system, rather than an extension of it.  Then I'd like to move our non-Cartesian coordinate systems into specialized math structures (or possibly just make contributions to crates such as cgmath)

Today we have the types Transform and LocalTransform, as well as a series of components for state management to keep the two in sync.  I say, why bother with clumsy synchronization?  We can combine the ideas.

```rust
#[derive(Clone)]
pub struct Transform {
    pub parent: Option<Entity>,
    pub matrix: [[f32; 4]; 4], 
}

impl Transform {
    // Convenience functions here for mutable and immutable borrows of x, y, z, 
    // and the ability to set and get rotation using either quaternions or euler angles
}

pub trait GlobalTransform {
    fn global_transform(&self, entity: Entity) -> [[f32; 4]; 4];
}

impl GlobalTransform for Storage<Transform> {
    fn global_transform(&self, entity: Entity) -> [[f32; 4]; 4] {
        // There may be some ownership issues in this function but I'm confident they can
        // be resolved.
        use cgmath::Matrix4;
        if let Some(transform) = self.get(entity) {
            if let Some(parent) = transform.parent {
                (Matrix4::from(self.get_global_transform(parent)) * Matrix4::from(self.matrix)).into()
            } else {
                self.matrix
            }
        }
    }
}
```

Then for PolarCoordinates we can provide a structure such as this one: https://docs.rs/nitro/0.9.1/nitro/math/struct.PolarCoords.html from my old engine Nitro.

**Comment by omni-viral** [comment]
How we are going to handle situation when entity has both `CartesianCoordinates` and `PolarCoordinates`?
Maybe make them part of enum instead?

**Comment by OvermindDL1** [comment]
> Maybe make them part of enum instead?

If a person wants to add 'yet another coordinate system' could they?  Does Rust have the concept of an open Enum like OCaml's Polymorphic Variants work?

**Comment by Binero** [comment]
@omni-viral We could do something like this: 

```rust
pub struct LocalTransform {
    pub position: Position,
    pub rotation: Rotation,
    pub scale: Vector3
}

pub enum Position {
    Cartesian(Vector3),
    Polar {
        radius: f32,
        angle: f32,
    }
}
```

Alternatively, we could simply have a `CartesianTransform` and a `PolarTransform`, and systems to sync them into `LocalTransform` when they are updated. Another system would then sync `LocalTransform` into `Transform` whenever `LocalTransform` is updated, or a parent's transform is updated. 

We could have a `GlobalTransform`, which would replace `LocalTransform`, `Child` and `Transform`. It would mean every new coordinate system would also need to re-implement the conversion step from `LocalTransform` to `Transform`, but this might not be much of an issue if there are public functions that can do that.

**Comment by Rhuagh** [comment]
My vote goes to having separate transform implementations for different coordinate system, will make it easier for users to add their own coordinate systems, and will also make it easier to interface with other libraries.

**Comment by omni-viral** [comment]
@OvermindDL1 enums in rust are not extensible easily. But there are not too many coordinate systems. We can enumerate them all. And if user want something exotic... maybe he makes a mistake 😄

**Comment by Xaeroxe** [comment]
So no one has any thoughts on the idea of collapsing the transform types?  I sort of just got completely ignored here.  Under my approach we wouldn't need any synchronizations and you could just convert between coordinate systems, do some calculations, then convert back to Cartesian.

**Comment by Binero** [comment]
@Xaeroxe I did post my thoughts about it. I would be in favor.

**Comment by Rhuagh** [comment]
@Xaeroxe I'd have to see how it works out in the renderer to have an opinion. Would the renderer call global_transform for all entities, in every frame or?

**Comment by Binero** [comment]
I would imagine a system where a `GlobalTransform` contains the following:

```rust
pub struct Transform {
    pub parent: Option<Entity>,
    pub local_matrix: [[f32; 4]; 4],
    pub matrix: [[f32; 4] 4],
}
```

If `parent.is_none()`, then `local_matrix == matrix`. Every time the parent updates or the `local_matrix` updates, `matrix` needs to be recalculated.

**Comment by Rhuagh** [comment]
Add a dirty flag and it might work.

**Comment by Rhuagh** [comment]
I know, it just avoids recomputation by sorting the entities.

**Comment by Binero** [comment]
In a scenario where an entity uses both `PolarTransform` and `CartesianTransform`, the last system would prevail. I think that's a fair behavior. An enum would limit ourselves into a limited set of coordinate systems, while a `Box<Transform>` would be very slow.

**Comment by Rhuagh** [comment]
I also don't want to implement polar coordinate calculation into my physics engine :P

**Comment by Xaeroxe** [comment]
@Rhuagh yeah there are a number of systems that we're only going to want Cartesian coordinates so conversions will be necessary.

**Comment by Rhuagh** [comment]
Having From implementations is easily done with `PolarTransform`, `CartesianTransform` etc.

**Comment by Binero** [comment]
Presumably those systems would use world-coordinates anyway, so they'd be using `GlobalTransform` rather than Cartesian or polar coordinates.

**Comment by Rhuagh** [comment]
For physics you want rotation and position separately though, you do _not_ want a single matrix.

**Comment by omni-viral** [comment]
@Rhuagh but in matrix rotation and position are in separate slots. Rotation in top-left 3x3 part and position in 4th column (or 4th row).

**Comment by Rhuagh** [comment]
If you ignore scaling yes.

**Comment by omni-viral** [comment]
@Rhuagh forgot about it 😄

**Comment by Xaeroxe** [comment]
Gitter log:

Jacob Kiesel
@Xaeroxe
13:53
Honestly I'm starting to think @Aceeri hit it on the nose when he wrote the Transform system as it exists today, and the only modification I'd like to make would be to provide conversion functions between [f32; 3] and a new PolarCoordinates structure which is not a component itself, or part of any components. That way you can convert, do your math, then convert back and store the results

I also might want to rename

Transform -> TransformMatrix
LocalTransform -> Transform
Jeroen Bollen
@Binero
13:54
That won't work as I interpret my rotation differently than in cartesian coordinates :confused:
in other words, it would still require a PolarTransform structure, and I would still have to sync it.
Jacob Kiesel
@Xaeroxe
13:55
So to clarify: Rotation as in the angle within the Polar coordinates, or rotation of the entity?
Jeroen Bollen
@Binero
13:56
Doesn't matter, I am happy syncing it as I planned
I still think GlobalTranform would be a more comprehensible interface though, then what we currently have (Init, Child, LocalTransform, InnerTransform & Transform)
Aceeri
@Aceeri
13:58
InnerTransform and Init are going away
Jeroen Bollen
@Binero
13:58
Okay, then I do agree we have no need for GlobalTransform

**Comment by OvermindDL1** [comment]
> @OvermindDL1 enums in rust are not extensible easily. But there are not too many coordinate systems. We can enumerate them all. And if user want something exotic... maybe he makes a mistake :smile:

Heh, I really *really* hate how many Haskell'isms rust takes, it should be taking from OCaml instead, the syntax for a lot of things in rust would be a lot more sane with first-class and functor modules for sure, plus hey Polymorphic Variants are almost trivial to implement on hardware. ^.^

But yeah, how about this for exotic, I implemented a sandbox once that was a 4d world and I did a 3d projection 'through' it based on various parameters.  ^.^;

(it was not efficient, but I was not trying to make it efficient, just... understandable)

**Comment by omni-viral** [comment]
@OvermindDL1 you rendered it onto 2d surface anyway 😏

**Comment by omni-viral** [comment]
> Polymorphic Variants are almost trivial to implement on hardware.

Can you explain it to someone who knows nothing about OCaml. Try to explain in C for example 😄

**Comment by Xaeroxe** [comment]
So our general conclusion seems to be Amethyst isn't the place for this abstraction to exist.  So, while this is technically unrelated to Amethyst I am going to say I have plans for adding an abstraction for polar and spherical coordinates to both mint and cgmath.  And with that I'm closing this issue.

**Comment by OvermindDL1** [comment]
> Can you explain it to someone who knows nothing about OCaml. Try to explain in C for example :smile:

Heh, a crash course in OCaml focused on Variants:

### Monomorphic Variants

A Monomorphic Variant is a variant that is bounded by a set amount of types and cannot be extended, can be like:

```ocaml
type t my_variant =
| Nothing
| Something of int
| Different of t * int
```
And this is a set of 3 heads of a variant/enum (I still don't know why rust/c calls it an enumeration, it's not like you enumerate through the heads...).  The `Nothing` head has no arguments, the `Something` head takes an integer, the `Different` head takes a whatever-you-specify type and an integer both.  Internally they are implemented in assembly as something like `Nothing` is a struct of (pseudo-code) `{tag, data}` where tag is an integer of 0 and data is sized the max data of all the heads here, but for 0 it is unused so can contain anything (in reality this first 'empty' case usually gets optimized out when used as a pointer to become the null pointer).  `Something` would have a tag of 1 with the first 31 bits storing an integer (OCaml integers are 31-bits by default, the last bit encodes if it is an integer or not, ignore this aspect though, and it usually gets optimized out anyway).  And `Different` will have a tag of 2 and it's data will store whatever the size of type `t` is followed by an integer, thus holding that data.  These can become decently unboxed at optimization time.

### Polymorphic Variants

```ocaml
`Nothing
`Something of int
`Different of whatever * int
```

Polymorphic variants are always boxed however, their type is either an integer or a tagged-structure.  At compile-time the signature and name of it become a unique hash in the system that is stored as either the integer that is a polymorphic variant at compile-time (if empty) or becomes the tag of the structure of the polymorphic variant at compile-time (if non-empty).  So \`Nothing would be an integer on the hardware of value 481346541 (it really is, I just ran it to check), and something like \`Something 42 would become the memory structure of the first integer-sized chunk of memory being `-681766822` and the next integer sized chunk of memory of the 2-integer sized chunk of allocated memory (it is boxed do remember) is `42`.  Something like \`Different (6.28, 42) would get a tag of `-890978471` with the data being a tuple (just a chunk of memory) sized of a float and an integer being `6.28, 42`.

Of course there are optimization passes that can pass it around on the stack directly (which tends to happen on functions that takes a set amount of polymorphic variants up to a certain size).

To use polymorphic variant's you just match on them like:

```ocaml
(* This accepts an open set of polymorphic variants, it takes these specific ones, but accepts any other too *)
let blah = function
| `Nothing -> "nothing"
| `Something i -> string_of_int i
| `Different (f, i) -> (string_of_float f) ^ " " ^ (string_of_int i)
| `DifferentUnbound (_, i) -> "unknown " ^ (string_of_int i)
| _ -> "unknown type"

(* And this is a closed set, it will not even compile if there is even a chance of a wrong one could get passed in*)
let bloop = function
| `Nothing -> "nothing"
| `Something i -> string_of_int i
| `Different (f, i) -> (string_of_float f) ^ " " ^ (string_of_int i)
```

All is of course *very* strongly typed (OCaml is full hindley-milner with First-class and higher-ordered module systems, which can represent anything haskell can do but more readably, though often more code, and it compiles *significantly* faster, I consider OCaml are far better, though less irritatingly pure, of a language than Haskell).

You can use it like:

```ocaml
> blah ~Nothing
"nothing"
> blah (~Something 42)
"42"
> blah (~Different (6.28, 42))
"6.28 42"
> blah (~DifferentUnbound ("vwoop", 42))
"unknown 42"
> blah ~Vwoop
"unknown type"

> bloop ~Nothing
"nothing"
> bloop (~Something 42)
"42"
> bloop (~Different (6.28, 42))
"6.28 42"
> bloop (~DifferentUnbound ("vwoop", 42))
"unknown 42"
> bloop ~Vwoop

Line 10, 10: This expression has type [> `Vwoop ]
       but an expression was expected of type
         [< `Different of float * int
          | `DifferentUnbound of `a * int
          | `Nothing
          | `Something of int ]
       The second variant type does not allow tag(s) `Vwoop
```

As you can see there are both open and closed polymorphic variant types (designated via `>` for open and `<` for closed.

A few points of note, OCaml averages at or near C speeds for equivalent code (it has a full optimizing compiler, one of the fastest out), it has magic to be able to break the type system if you really really need (with the usual caveats and you really need to know what things compile to on the metal) but otherwise it is *very* strongly typed, it has records, both variant types as you saw above, GADT's, row-typed records (think kind of like classes but yet not, these are often not able to be optimized as well and incur some function pointer indirection, just like usual classes do, OCaml often compiles everything else *really* well, even functor modules often get inlined and optimized for their types), first-class modules (you can pack up a namespace/module with a signature and unpack it to use it elsewhere, great to use as Witnesses!), functor modules (you can have modules that operate over modules to generate new modules (or more functors), think templates in C++ but reigned in and static at run-time (this all happens at compile-time)), among quite a bit more.  It lacks multi-threading work (it has a GIL currently) but there is a big PR that is adding full multi-thread capability, no GIL, with/via Algebraic Effects soonish (you can implement any form of concurrent that you can think of with it).  OCaml also supports compile-time language translation (you can make a PP layer that translates one text file input into OCaml source) and compile-time ast translation (you can make a PPX layer at any point along the pipeline of the AST compiling and optimization stages so you can do your own source translations, add new features, etc...)

Plus I love the compiling times of a few seconds on average compared to C++/Rust/Haskell's ocraphelpmethisistakingforever times for *very* complex programs (you'd average <1s compiling time for most programs).

Rust took a lot of it's coding and style from both Haskell and OCaml worlds, sadly it took too much from Haskell and not enough from OCaml, which further contributes to it's dog-slow compiling time and some really (in my opinion) quite stupid decisions in some of the syntax that contributes to it's slow compiling times.  OCaml's Module system is the most powerful in the Module-Language world, and as such it can emulate just about any language feature of Haskell, while being faster to execute and compiler both.  OCaml also has an upcoming 'Implicit Module' PR sometime that will allow for auto-selection of in-scope modules based on types (basically implicit witnesses, think Haskell's type-classes but *significantly* more powerful, like having multiple types to match on as one example of many, and overrideable based on scope as well).

Basically if you know Haskell, OCaml is it's (older) cousin that actually does the type system Correctly, designed for real-world research (backwards compat is important, compiling time speed is important, runtime speed is important, being able to break the system if absolute necessary is important (and is all gated behind the mystical `Obj.magic` function, think of it like Rusts's `unsafe` keyword, except it's a function, and it's only power is basically `reinterpret_cast<t>` from the C++ world).
